cmd/cosign/cli/dockerfile/resolve.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 The Sigstore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dockerfile
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/sigstore/cosign/pkg/oci/remote"
+
+	"github.com/pkg/errors"
+)
+
+// ResolveDockerfileCommand rewrites the Dockerfile
+// base images to FROM <digest>.
+type ResolveDockerfileCommand struct {
+	Output string
+}
+
+// Exec runs the resolve dockerfile command
+func (c *ResolveDockerfileCommand) Exec(ctx context.Context, args []string) error {
+	if len(args) != 1 {
+		return flag.ErrHelp
+	}
+
+	dockerfile, err := os.Open(args[0])
+	if err != nil {
+		return fmt.Errorf("could not open Dockerfile: %w", err)
+	}
+	defer dockerfile.Close()
+
+	resolvedDockerfile, err := resolveDigest(dockerfile)
+	if err != nil {
+		return fmt.Errorf("failed extracting images from Dockerfile: %w", err)
+	}
+
+	if c.Output != "" {
+		if err := os.WriteFile(c.Output, resolvedDockerfile, 0600); err != nil {
+			return fmt.Errorf("failed writing resolved Dockerfile: %w", err)
+		}
+	} else {
+		fmt.Fprintln(os.Stdout, string(resolvedDockerfile))
+	}
+
+	return nil
+}
+
+func resolveDigest(dockerfile io.Reader) ([]byte, error) {
+	fileScanner := bufio.NewScanner(dockerfile)
+	tmp := bytes.NewBuffer([]byte{})
+
+	for fileScanner.Scan() {
+		line := strings.TrimSpace(fileScanner.Text())
+
+		if strings.HasPrefix(strings.ToUpper(line), "FROM") ||
+			strings.HasPrefix(strings.ToUpper(line), "COPY") {
+			switch image := getImageFromLine(line); image {
+			case "scratch":
+				tmp.WriteString(line)
+			default:
+				ref, err := name.ParseReference(image)
+				if err != nil {
+					// we should not return err here since
+					// we can define the image refs smth like
+					// i.e., FROM alpine:$(TAG), FROM $(IMAGE), etc.
+					// TODO: support parameter substitution by passing a `--build-arg` flag
+					fmt.Fprintf(os.Stderr, "WARNING: parameter substitution for images does not supported yet. consider set all environment variables before resolving the Dockerfile. Image: %s.\n", image)
```

Comment:
```suggestion
					fmt.Fprintf(os.Stderr, "WARNING: parameter substitution for images is not supported yet. consider set all environment variables before resolving the Dockerfile. Image: %s.\n", image)
```

I would also like to consider that this warning is made a little more clear. I don't have context, but maybe this warning could be slightly better written for usability 😄

Reply:
Also, surely this is not the only scenario where this function could fail?

Reply:
Thanks!

> only scenario where this function could fail

in case if we couldn't parse the image reference